tests/storage/databases/main/test_events_worker.py

```diff
@@ -121,6 +132,42 @@ def test_query_via_event_cache(self):
             self.assertEqual(res, {self.event_ids[0]})
             self.assertEqual(ctx.get_resource_usage().db_txn_count, 0)
 
+    def test_persisting_event_invalidates_cache(self):
+        with LoggingContext(name="test") as ctx:
+            alice = self.register_user("alice", "pass")
+            alice_token = self.login("alice", "pass")
+            room_id = self.helper.create_room_as(alice, tok=alice_token)
```

Comment:
Just trying to get a local reproduction unit test for this. But this simple setup runs into `psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "events_stream_ordering"
	DETAIL:  Key (stream_ordering)=(2) already exists.` when creating the room here
	
---

```
SYNAPSE_POSTGRES=1 SYNAPSE_TEST_LOG_LEVEL=INFO python -m twisted.trial tests.storage.databases.main.test_events_worker.HaveSeenEventsTestCase.test_persisting_event_invalidates_cache

tests.storage.databases.main.test_events_worker
  HaveSeenEventsTestCase
    test_persisting_event_invalidates_cache ...                         [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/tests/storage/databases/main/test_events_worker.py", line 139, in test_persisting_event_invalidates_cache
    room_id = self.helper.create_room_as(alice, tok=alice_token)
  File "/Users/eric/Documents/github/element/synapse/tests/rest/client/utils.py", line 143, in create_room_as
    assert channel.code == expect_code, channel.result
builtins.AssertionError: {'version': b'1.1', 'code': b'500', 'reason': b'Internal Server Error', 'headers': [(b'Server', b'1'), (b'Date', b'Wed, 21 Sep 2022 18:36:23 GMT'), (b'Synapse-Trace-Id', b'0'), (b'Content-Type', b'application/json'), (b'Cache-Control', b'no-cache, no-store, must-revalidate'), (b'Access-Control-Allow-Origin', b'*'), (b'Access-Control-Allow-Methods', b'GET, HEAD, POST, PUT, DELETE, OPTIONS'), (b'Access-Control-Allow-Headers', b'X-Requested-With, Content-Type, Authorization, Date')], 'body': b'{"errcode":"M_UNKNOWN","error":"Internal server error"}', 'done': True}
```

And `test.log` shows `psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "events_stream_ordering"
	DETAIL:  Key (stream_ordering)=(2) already exists.`:

<details>
<summary>Full error</summary>

```
2022-09-21 02:18:28-0500 [-] 2022-09-21 02:18:28,416 - synapse.http.server - 123 - ERROR - POST-3 - Failed handle request via 'RoomCreateRestServlet': <SynapseRequest at 0x114798a00 method='POST' uri='/_matrix/client/r0/createRoom?access_token=<redacted>' clientproto='1.1' site='test'>
	Traceback (most recent call last):
	  File "synapse/synapse/http/server.py", line 306, in _async_render_wrapper
	    callback_return = await self._async_render(request)
	  File "synapse/synapse/http/server.py", line 512, in _async_render
	    callback_return = await raw_callback_return
	  File "synapse/synapse/rest/client/room.py", line 161, in on_POST
	    info, _ = await self._room_creation_handler.create_room(
	  File "synapse/synapse/handlers/room.py", line 909, in create_room
	    ) = await self._send_events_for_new_room(
	  File "synapse/synapse/handlers/room.py", line 1105, in _send_events_for_new_room
	    await send(etype=EventTypes.Create, content=creation_content)
	  File "synapse/synapse/handlers/room.py", line 1081, in send
	    ) = await self.event_creation_handler.create_and_send_nonmember_event(
	  File "synapse/synapse/handlers/message.py", line 990, in create_and_send_nonmember_event
	    ev = await self.handle_new_client_event(
	  File "synapse/synapse/util/metrics.py", line 113, in measured_func
	    r = await func(self, *args, **kwargs)
	  File "synapse/synapse/handlers/message.py", line 1325, in handle_new_client_event
	    result, _ = await make_deferred_yieldable(
	  File "synapse/synapse/storage/controllers/persist_events.py", line 287, in handle_queue_loop
	    ret = await self._per_item_callback(room_id, item.task)
	  File "synapse/synapse/storage/controllers/persist_events.py", line 359, in _process_event_persist_queue_task
	    return await self._persist_event_batch(room_id, task)
	  File "synapse/synapse/storage/controllers/persist_events.py", line 748, in _persist_event_batch
	    await self.persist_events_store._persist_events_and_state_updates(
	  File "synapse/synapse/logging/tracing.py", line 851, in _wrapper
	    return await func(*args, **kwargs)  # type: ignore[misc]
	  File "synapse/synapse/storage/databases/main/events.py", line 214, in _persist_events_and_state_updates
	    await self.db_pool.runInteraction(
	  File "synapse/synapse/storage/database.py", line 881, in runInteraction
	    return await delay_cancellation(_runInteraction())
	  File "synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1656, in _inlineCallbacks
	    result = current_context.run(
	  File "synapse/.venv/lib/python3.9/site-packages/twisted/python/failure.py", line 514, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "synapse/synapse/storage/database.py", line 848, in _runInteraction
	    result = await self.runWithConnection(
	  File "synapse/synapse/storage/database.py", line 972, in runWithConnection
	    return await make_deferred_yieldable(
	  File "synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 857, in _runCallbacks
	    current.result = callback(  # type: ignore[misc]
	  File "synapse/tests/server.py", line 508, in <lambda>
	    d.addCallback(lambda x: function(*args, **kwargs))
	  File "synapse/.venv/lib/python3.9/site-packages/twisted/enterprise/adbapi.py", line 282, in _runWithConnection
	    result = func(conn, *args, **kw)
	  File "synapse/synapse/storage/database.py", line 965, in inner_func
	    return func(db_conn, *args, **kwargs)
	  File "synapse/synapse/storage/database.py", line 710, in new_transaction
	    r = func(cursor, *args, **kwargs)
	  File "synapse/synapse/storage/databases/main/events.py", line 435, in _persist_events_txn
	    self._store_event_txn(txn, events_and_contexts=events_and_contexts)
	  File "synapse/synapse/storage/databases/main/events.py", line 1459, in _store_event_txn
	    self.db_pool.simple_insert_many_txn(
	  File "synapse/synapse/storage/database.py", line 1111, in simple_insert_many_txn
	    txn.execute_values(sql, values, fetch=False)
	  File "synapse/synapse/storage/database.py", line 383, in execute_values
	    return self._do_execute(
	  File "synapse/synapse/storage/database.py", line 432, in _do_execute
	    return func(sql, *args, **kwargs)
	  File "synapse/synapse/storage/database.py", line 384, in <lambda>
	    lambda the_sql: execute_values(self.txn, the_sql, values, fetch=fetch),
	  File "synapse/.venv/lib/python3.9/site-packages/psycopg2/extras.py", line 1270, in execute_values
	    cur.execute(b''.join(parts))
	psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "events_stream_ordering"
	DETAIL:  Key (stream_ordering)=(2) already exists.
```

</details>

Reply:
It looks like it's the manual `events` and `event_json` inserting we do in the `prepare` and not incrementing the stream ordering.

https://github.com/matrix-org/synapse/blob/efabf44c7652095a0e3d9d9083fc8359cdde3854/tests/storage/databases/main/test_events_worker.py#L41-L92